### ISSUE TITLE
scanf/printf  is/are broken

### ISSUE THREAD

**Issue by mheyer32** [bebbo/libnix#19]
Likely scanf is the offender as ScummVM seems not to be able to parse the ini file correctly,

This was introduced with commit 396a8e85029ecfd158a3ef0f1af492cb3a958bae

Maybe I can dig a bit deeper tomorrow.

**Comment by arczi84** [comment]
Same with sprintf.
But snprintf works good.

**Comment by bebbo** [comment]
These are the current results for libc-test:

Test | newlib | libnix | clib2 | gcc/cygwin
:--- | :---: | :---: | :---: | :---:
sscanf |  3 |  0 |  0(39) |  3
snprintf | 0 |  0 |  0(39) |  1

**Comment by mheyer32** [comment]
Have you tried running the test with (winuae)enforcer?
Maybe the functions themselves are mostly ok but write past buffer ends?

**Comment by mheyer32** [comment]
Its still not working.
I'm looking at a suspect here:
https://github.com/bebbo/libnix/commit/396a8e85029ecfd158a3ef0f1af492cb3a958bae#r32536252

There's also a similar minor change that would limit the exponent in a floating point number to 100 - but I don't think this is relevant.

**Comment by bebbo** [comment]
I'm still missing the info what's broken...

 - scummvm reads the ini and lists the game [sq1sci].
 - SQ1 loads and seems to playable (I'm not good at playing it yet)
 - no enforcer hits
 - no winuaeenforcer hits
 - `0x` with an invalid numer must not continue parsing, the break exits the scanf loop, whilest a missing `x` is valid.
 - the exponent is not limited to 99, it's valid to provide any exponent.

**Comment by mheyer32** [comment]
So what happens here:
DiscWorld will not load, claiming there's no game data
SC1 will start, load for a while and then crash with a std::bad_alloc(), followed by the TimerManager thread crashing.

(The TimerManager crash is probably just a bad side effect, not an actual problem. I believe its because the main program aborts(), its code and data goes out the window and this pulls the rug under the TimerManager process. Maybe I can install some form of abort() handler and try tearing down the extra processes there?).

Try the following. Configure to 'full debug'
`./configure --host=m68k-amigaos --disable-all-engines --enable-engine=tinsel,sci --disable-mt32emu --enable-debug --disable-optimizations --disable-hq-scalers --with-amiga-prefix=/media/sf_Amiga/ScummVM --disable-translation --enable-c++11  --disable-updates`

Force the debug output level to something high:
```
OSystem_AmigaOS3::OSystem_AmigaOS3() {
	 gDebugLevel = 11;
```

Add some debug output to  Scumm's  vformat()
```
String String::vformat(const char *fmt, va_list args) {


...
       		// vsnprintf didn't have enough space, so grow buffer
		printf("String::vformat allocates %d bytes for format \"%s\"\n", len, fmt);
		output.ensureCapacity(len, false);
}
```

Then ScummVM  will not even start properly and immediately abort with bad_alloc :-)

![image](https://user-images.githubusercontent.com/33539855/53623823-0de7f180-3bb3-11e9-888c-90692411776f.png)

**Comment by mheyer32** [comment]
It looks like vsnprintf() doesn't like it when there's no actual format thingies in the format string:

The below program prints a pretty large number :-)

```
static void myPrint(const char* fmt) {
   char str[32];
   int len = vsnprintf(str, sizeof(str), fmt, NULL);
   printf("%d", len); fflush(stdout);
}

int main(int argc, char ** argv) {
    myPrint("hjjjjjjjjjjjashjashjkahsjkahslajhsjkahsjlkhslajhsajklhsajklhs");
  return 0;
}
```

**Comment by bebbo** [comment]
> followed by the TimerManager thread crashing.

guess there is no atexit() hanling which shuts it properly down

**Comment by bebbo** [comment]
> It looks like vsnprintf() doesn't like it when there's no actual format thingies in the format string:
> The below program prints a pretty large number :-)

61 is not that large!?

**Comment by bebbo** [comment]
do you still use features like `-fshort-enums` to build your lib or so?

**Comment by bebbo** [comment]
> 
> 
> > followed by the TimerManager thread crashing.
> 
> guess there is no atexit() handling which shuts it properly down

```
diff --git a/engines/sci/graphics/palette.cpp b/engines/sci/graphics/palette.cpp
index bc0b8485d2..821b5c6c29 100644
:
+++ b/backends/platform/amigaos3/amigaos3-main.cpp
@@ -171,6 +171,18 @@ static void load_libraries(void) {
        }
 }

+void cleanup() {
+       // Delete OSystem
+       if (g_system) {
+               delete (OSystem_AmigaOS3 *)g_system;
+       }
+
+       if (wbClosed) {
+               OpenWorkBench();
+       }
+
+}
+
 __stdargs int main(int argcWb, char const * argvWb[]) {
        load_libraries();

@@ -182,10 +194,12 @@ __stdargs int main(int argcWb, char const * argvWb[]) {
        int audioThreadPriority = DEFAULT_AUDIO_THREAD_PRIORITY;
        int closeWb = 0;

-       struct Process* proc = (struct Process*)FindTask(NULL);
-       printf("Process stack size is %ld bytes.\n", proc->pr_StackSize);
-       if (proc->pr_StackSize < 100000) {
-               printf("Process stack size is smaller than 100000. Exiting.\n");
+       struct Task * task = FindTask(NULL);
+       unsigned ss = (char*)task->tc_SPUpper - (char*)task->tc_SPLower;
+       struct Process* proc = (struct Process*)task;
+//     printf("Process stack size is %ld bytes.\n", ss);
+       if (ss < 100000) {
+               printf("Process stack size is smaller than 100000. Exiting. %d\n", ss);
                exit(EXIT_FAILURE);
        }

@@ -218,6 +232,7 @@ __stdargs int main(int argcWb, char const * argvWb[]) {
        OSystem_AmigaOS3 *amigaOsSystem = new OSystem_AmigaOS3();
        g_system = amigaOsSystem;
        assert(g_system);
+       atexit(cleanup);

        // Pre initialize the backend
        amigaOsSystem->init(audioThreadPriority);
@@ -226,17 +241,9 @@ __stdargs int main(int argcWb, char const * argvWb[]) {
                wbClosed = CloseWorkBench();
        }

+
        // Invoke the actual ScummVM main entry point:
        int res = scummvm_main(argc, argv);

-       // Delete OSystem
-       if (g_system) {
-               delete amigaOsSystem;
-       }
-
-       if (wbClosed) {
-               OpenWorkBench();
-       }
-
        return res;
 }
```

**Comment by mheyer32** [comment]
> do you still use features like `-fshort-enums` to build your lib or so?

No, I tried with and without, both result in the same output. So for you above example did produce correct results?

I just did  a clean, clean-prefix, all, all-sdk rebuild and the example code above is still producing a weird number.

**Comment by bebbo** [comment]
I only added the missing `#include` to your code
```
> m68k-amigaos-gcc -Os vpr.c -o vpr -noixemul
>  vamos vpr
61
```
also compiled it with the downloadable setup-amiga-gcc.exe package
=> 61

... will test running Linux soon

**Comment by bebbo** [comment]
Linux compiled binary also prints 61...
Please attach your binary.

**Comment by mheyer32** [comment]
Not sure if it makes a difference, but try compiling with g++  as a cpp file. 
Thats what I did and ScummVM is also compiled as C++ project.


[Test.zip](https://github.com/bebbo/libnix/files/2920929/Test.zip)

**Comment by bebbo** [comment]
perdon? you compiled this 
[vpr.zip](https://github.com/bebbo/libnix/files/2920986/vpr.zip)
with `m68k-amigaos-g++ -Os vpr.c -o vpr -noixemul`?

**Comment by bebbo** [comment]
printf isn't broken - please tell me your CFLAGS_FOR_TARGET.

**Comment by mheyer32** [comment]
I can't do it right now, I'll follow up tomorrow.

Can you try compiling vpr.c as .cpp file?

I'll triple-check that I'm not messing with CFLAGS_FOR_TARGET... Or better yet,  delete and rebuilt the whole toochain directory.

**Comment by bebbo** [comment]
check your .bashrc etc.p.p

I rather suspect a further bug in my optimizer.

**Comment by bebbo** [comment]
my guess: you are using `-fomit-frame-pointer -Ofast`
